### PR TITLE
Train frontend_server snapshot

### PR DESCRIPTION
**Pull request by jensjoha**
When running with --preview-dart-2 the frontend_server is used for compilation.
The frontend_server is compiled to a snapshot and trained with "--train", but then contains the following code "if (options['train']) { return 0; }" meaning that it isn't trained at all.

This PR trains it on itself resulting in the following improvements in the "hot_mode_dev_cycle" "hotReloadInitialDevFSSyncMilliseconds" benchmark:

```
On x64 Android emulator:

dart bin/run.dart -t hot_mode_dev_cycle__preview_dart_2_benchmark
hotReloadInitialDevFSSyncMilliseconds goes from an average (3 runs) of 5905 to 4178

dart bin/run.dart -t hot_mode_dev_cycle__benchmark
hotReloadInitialDevFSSyncMilliseconds takes an average (3 runs) of 5375.



On Nexus 4 running Android 4.4.4:

dart bin/run.dart -t hot_mode_dev_cycle__preview_dart_2_benchmark
hotReloadInitialDevFSSyncMilliseconds goes from an average (3 runs) of 8493 to 6863

dart bin/run.dart -t hot_mode_dev_cycle__benchmark
hotReloadInitialDevFSSyncMilliseconds takes an average (3 runs) of 9218.
```

### CODE REVIEW

**Comment by jensjoha** [comment]
@aam and/or @mraleph would you review this change?

File: frontend_server/lib/server.dart [hunk 1]
```diff
@@ -259,6 +259,23 @@ Future<int> starter(
   }
 
   if (options['train']) {
+    String sdkRoot = options['sdk-root'];
+    if (sdkRoot.startsWith("//out/") && sdkRoot.endsWith("/gen")) {
+      // This becomes '../../out/whatnot/flutter_patched_sdk'
+      sdkRoot = '../..${sdkRoot.substring(1, sdkRoot.length-4)}/flutter_patched_sdk';
```
Comment (mraleph) [inline frontend_server/lib/server.dart]: This is really hacky, I don't like it. Is there no way to make GN properly rebase paths and get absolute path in here?

Maybe @zanderso knows the trick?

**Comment by mraleph** [comment]
I think this looks good to me, sans my comment.

Do you know how long does the training take?

**Comment by jensjoha** [comment]
~5 seconds on my machine...

File: frontend_server/lib/server.dart (hunk 1 above)
Comment (jensjoha) [reply frontend_server/lib/server.dart]: How that you implicitly told me what this rebase_path command I've seen used does, then yes. Fixed.

**Comment by mraleph** [comment]
LGTM.

Lets see if @aam has any reservations about this and if he does not we will merge.

**Comment by aam** [comment]
LGTM too! Thank you for fixing this.

File: frontend_server/BUILD.gn [hunk 2]
```diff
@@ -7,7 +7,8 @@ import("//third_party/dart/utils/application_snapshot.gni")
 application_snapshot("frontend_server") {
   main_dart = "bin/starter.dart"
   dot_packages = rebase_path(".packages")
-  training_args = [ "--train" ]
+  flutter_patched_sdk = rebase_path("$root_gen_dir/../flutter_patched_sdk")
```
[the remark this answers is not in the document]
Comment (jensjoha) [reply frontend_server/BUILD.gn]: You appear to be correct. Changed.

**Comment by mraleph** [comment]
@jensjoha thanks for the update. I will merge as soon as Travis cycles.

**Comment by aam** [comment]
This seem to have broken flutter/engine Windows build https://build.chromium.org/p/client.flutter/console

**Pull request merged**